model-optimizer/extensions/front/kaldi/replace_timeheightconvolution.py

```diff
@@ -0,0 +1,122 @@
+                memoff.in_port(0).connect(node.in_port(0).get_source())
+                i = i + 1
+            prev = concat.out_port(0)
+        # no need to create concat for 1 offset
+        elif len(all_time_offsets) == 1:
+            t = all_time_offsets.pop()
+            has_default = t not in req_time_offsets
+            memoff = MemoryOffset(graph, attrs={'name': in_name + '/MemoryOffset',
+                                                'has_default': has_default, 'splitted': False,
+                                                'pair_name': in_name + '/MemoryOffset_pair',
+                                                't': t}).create_node()
+            memoff.in_port(0).connect(node.in_port(0).get_source())
+            prev = memoff.out_port(0)
+
+        stride = node.soft_get("height_subsample", 1)
+
+        kernel = int64_array([0, 0])
+        kernel[0] = len(set(offsets[:, 0]))
+        kernel[1] = len(set(offsets[:, 1]))
+
+        pad_h = int64_array([0, 0])
+        pad_h[0] = -min(offsets[:, 1]) if min(offsets[:, 1]) < 0 else 0
+        pad_h[1] = stride * node.height_out - (node.height_in - max([max(offsets[:, 1]), 0]))
+
+        dilation_t = (max(offsets[:, 0]) - min(offsets[:, 0])) / (kernel[0] - 1) if kernel[0] > 1 else 1
+        dilation_h = (max(offsets[:, 1]) - min(offsets[:, 1])) / (kernel[1] - 1) if kernel[0] > 1 else 1
+
+        mapping_rule = {
+            'name': in_name,
+            'output': node['out_channels'],
+            'patch_stride': node.height_in,
+            'bias_term': None,
+            'pad': int64_array([[0, 0], [0, 0], [0, 0], pad_h]),
+            'pad_spatial_shape': int64_array([[0, 0], pad_h]),
+            'dilation': int64_array([1, 1, dilation_t, dilation_h]),
+            'kernel': int64_array([node.out_channels, node.in_channels, kernel[0], kernel[1]]),
+            'stride': int64_array([1, 1, 1, stride]),
+            'kernel_spatial': kernel,
+            'input_feature_channel': 1,
+            'output_feature_channel': 0,
+            'channel_dims': int64_array([1]),
+            'spatial_dims': int64_array([2, 3]),
+            'batch_dims': int64_array([0]),
+            'kernel_spatial_idx': int64_array([2, 3]),
+            'group': 1,
+            'reshape_kernel': True,
+            'bias_addable': True,
+        }
+        conv = Convolution(graph, attrs=mapping_rule).create_node()
+        conv.in_port(0).connect(prev)
+        conv.in_port(1).connect(node.in_port(1).get_source())
+
+        # change layout for weights from OHWI to OIHW
+        # in future should be replaced by common Permute mechanics
+        weights = conv.in_port(1).get_source().node.value
+        weights = weights.reshape(int64_array([node.out_channels, -1, node.in_channels]))
+        weights = weights.transpose(int64_array([0, 2, 1]))
+        weights = weights.flatten()
+        conv.in_port(1).get_source().node.value = weights
```

Comment:
nit: It is better to avoid such a Const node manipulations. Instead, build a graph with `Reshape->Transpose->Flatten`. The constant folding will do the rest for you. I understand that for Kaldi models the weights are always constants, but so we may keep the current implementation as is.

Reply:
let's keep